src/DomainServices/DocumentService.ts

```diff
@@ -54,24 +60,101 @@ export class DocumentService {
   }
 
   public async validateUserAccess(
-    user: Express.User,
+    userID: string,
     projectID: string,
     permission: DocumentPermission
   ) {
-    const permissions = await this.getPermissions(projectID, user.id)
+    const permissions = await this.getPermissions(projectID, userID)
     if (!permissions.has(permission)) {
-      throw new RoleDoesNotPermitOperationError(`Access denied`, user.id)
+      throw new RoleDoesNotPermitOperationError(`Access denied`, userID)
     }
   }
 
+  public async validateTokenAccess(
+    token: string,
+    projectID: string,
+    permission: DocumentPermission
+  ) {
+    const { userID } = validateToken(token)
+    await this.validateUserAccess(userID, projectID, permission)
+  }
+
   public async getManuscriptFromSnapshot(snapshot: Snapshot) {
     const manuscriptID = snapshot.doc_id
-    const manuscript: Manuscript | null =
-      await DIContainer.sharedContainer.projectClient.getProject(manuscriptID)
+    const manuscript = await DIContainer.sharedContainer.projectClient.getProject(manuscriptID)
 
     if (!manuscript) {
       throw new MissingManuscriptError(manuscriptID)
     }
     return manuscript
   }
+
+  public async handleUpgrade(
+    wss: WebSocketServer,
+    request: IncomingMessage,
+    socket: Duplex,
+    head: Buffer
+  ) {
+    const url = request.url
+    if (!url) {
+      this.destroyStream(socket)
+      return
+    }
+
+    const { manuscriptID, projectID } = this.parseConnectionURL(url)
+    if (!manuscriptID || !projectID) {
+      this.destroyStream(socket)
+      return
+    }
+
+    wss.handleUpgrade(request, socket, head, (ws) => {
+      this.attachListeners(socket, ws, manuscriptID)
+      this.socketsService.setClient(manuscriptID, ws)
+      console.log('client has been set')
+      wss.emit('connection', ws, request)
+    })
+  }
+  private parseConnectionURL(url: string) {
+    const urlRegex =
+      /^\/api\/v2\/doc\/(MPProject:[0-9a-fA-F-]+)\/manuscript\/(MPManuscript:[0-9a-fA-F-]+)\/listen/
+    const match = url.match(urlRegex)
+    return {
+      projectID: match ? match[1] : undefined,
+      manuscriptID: match ? match[2] : undefined,
+    }
+  }
+  private attachListeners(_stream: Duplex, ws: WebSocket, manuscriptID: string) {
+    ws.onerror = (error) => this.onError(ws, error)
```

Comment:
I don't see an `onmessage` handler here, presumably because the FE currently doesn't send the info using it.

Reply:
Yes, that's correct.